### ISSUE TITLE
Signing in causes me to lose my work.

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.webmaker.org#627]
Steps to reproduce:
* Go to https://bramble.mofostaging.net/
* Make some edits to the starter project
* Click Sign In
* Complete the Sign In process
* You end up on the **Your Projects** view
* You've lost the changes you made prior to Signing In

You should end up in the editor, with the same content you had before signing in.

**Suggested Flow**
When someone signs in or creates an account, we should save whatever work they had in the editor as a new project and return the user there after they complete the sign up/in process so they can continue editing.

**Comment by humphd** [comment]
cc @gideonthomas, @sedge

**Comment by flukeout** [comment]
So does refreshing the browser while I'm signed in.

**Comment by gideonthomas** [comment]
So, if you do some work in thimble before signing in and then sign in, we don't have any UI/UX to alert the user whether they want to discard existing changes or to create a new project out of those changes, etc. This is part of the anonymous user's workflow which we haven't implemented yet. Hence, for now, you lose all those changes and you'll be shown the "Your Projects" page.

As far as refreshing the browser goes, we had it always show you the "Your Projects" page (knowing that we would have to change it eventually), but I can fix that right now so that it opens up whichever project you previously had!

**Comment by gideonthomas** [comment]
Part of this fix is [here](https://github.com/mozilla/thimble.webmaker.org/pull/640)

**Comment by flukeout** [comment]
Updated the ticket description proposed a solution.

**Comment by flukeout** [comment]
ping @gideonthomas - how is progress on this one?

**Comment by sedge** [comment]
@flukeout I'm going to take this on

**Comment by sedge** [comment]
This is blocked on #570

**Comment by gideonthomas** [comment]
This is now blocked on https://github.com/mozilla/thimble.webmaker.org/issues/756

**Comment by sedge** [comment]
@humphd and I did some thinking on this today. We figure there are two options, both of which hinge on storing anonymous projects in a special folder:

1. When a signed-in user opens thimble, and there are anonymous projects in the folder, open a dialog telling the user that there are unsaved projects. The user would be able to save each project from here, and when they did so, the anonymous project would be deleted after it had been upgraded. This isn't ideal because of the huge amount of UX/UI work that would be needed to get it running. 
2. We make the anonymous user flow stateless. 

Our raw notes for number 2 are on [this etherpad](https://etherpad.mozilla.org/thimbleanon), but here's the summary:

***Unauthenticated project flow***
*A user accesses thimble.com, either directly or through `/remix`*

* The server generates a UUID and redirects the user to a new route: `/:uuid/:remixid`
* The server renders the thimble homepage, including the `uuid`, `remixid` and project title somewhere on the page
* The client first checks the local filesystem to see if `/.anon/projects/<uuid>` exists.
  * If it does, it mounts that directory and opens the project.
  * If it doesn't, it creates that folder and calls `/initializeProject` passing `remixid` as a querystring param if it existed: `/initializeProject?...&remixid=123` or `/initializeProject?...`
* The server sends the appropriate files to the client
  * If `remixid` was passed, it collects the remix files and returns those
  * If it didn't, it returns the default project files
* The client populates the folder `/.anon/projects/<uuid>` with the files and mounts the folder

***Authenticated project flow***
*A user has bookmarked the anon-project url and returns to it after signing in, or refreshes the page after signing in in a different tab*

* The client hits a thimble route to create a project, and then checks the local filesystem to see if `/.anon/projects/<uuid>` exists.
  * If it doesn't, it hits `/initializeProject` like it does now, populating the project with the default content, saving and syncing along the way.
  * If it does, it copies the files to the new project's root and save and syncs all.

***Notes:***
* This effectively removes server state for unauthenticated users. 
* This relies on the user keeping the url in order to access their project again. We could have UX for this
* This is browser exclusive, meaning an anonymous project in one browser can't be accessed from another even with the same url

In conclusion, this solves many of the problems we have with the unauthenticated user flow. I'm not convinced that it's perfect, since the risk of losing the url and thus the anonymous project is high. I suspect we'll need some UI one way or another to show what a user has in the `/.anon` folder.

**Comment by sedge** [comment]
@gideonthomas Let's call the uuid `anonymousId` when it's rendered into the thimble home page.

**Comment by gideonthomas** [comment]
@sedge the query string for the login url should look something like: whatever it already has + `&title=A&anonymousId=1&now=111111`

**Comment by sedge** [comment]
@gideonthomas The querystring login no longer needs the title correct?

**Comment by gideonthomas** [comment]
@sedge it still needs the title
